compiler/rustc_lint_defs/src/builtin.rs

```diff
@@ -3272,6 +3272,44 @@ declare_lint! {
     "ambiguous glob re-exports",
 }
 
+declare_lint! {
+    /// The `local_binding_shadows_glob_reexport` lint detects cases where a glob re-exported
+    /// name would be shadowed by a local binding such as a new type declaration or an import
+    /// statement.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![deny(local_binding_shadows_glob_reexport)]
+    ///
+    /// pub mod upstream {
+    ///     mod inner { pub struct Foo {}; pub struct Bar {}; }
+    ///     pub use self::inner::*;
+    ///     struct Foo {} // local binding shadows `inner::Foo`
+    /// }
+    ///
+    /// // mod downstream {
+    /// //     fn test() {
+    /// //         let _ = crate::upstream::Foo; // inaccessible
+    /// //     }
+    /// // }
+    ///
+    /// pub fn main() {}
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// This was previously accepted without any errors or warnings but it could silently break a
+    /// crate's downstream user code. If the `struct Foo` was added, `dep::inner::Foo` would
+    /// silently become inaccessible and trigger a "`struct `Foo` is private`" visibility error in
+    /// the downstream use site.
+    pub LOCAL_BINDING_SHADOWS_GLOB_REEXPORT,
```

Comment:
```suggestion
    pub HIDDEN_GLOB_REEXPORTS,
```
Let's use a bit more general name.

In retrospect I'd probably merge this lint and `ambiguous_glob_reexports` into a single lint.
In both cases we have a public glob reexport and we think that it reexports everything public in its target module, but it actually does not due to inference from other items.

Also "binding" in this context is an internal jargon, in public docs we should probably say that the glob reexport is shadowed by a private item (this applies to all public facing text - docs, error messages).